2019-12-naming-part1/naming.md

```diff
@@ -0,0 +1,231 @@
+  end
+end
+```
+
+At this point, it might be tempting to reach for an overtly literal function name, like `split_and_join` above. Indeed, the code splits and joins a name. It's quick and easy to conceive of; it's convenient.
+
+The name `split_and_join` mirrors the function's internals. Refer back to the Abstraction Ladder diagram above. Linguistically, Molly seems like concrete entity, but internally she's a collection of subprocesses. It would be technically correct, but awfully strange, to have named Molly after her internal subprocesses, perhaps something like `organs_and_bones_and_waggy_tail`.
+
+Names should establish an identity that's relatively unique. It's conceivable that other dogs in the world will be named Molly, but it's unlikely that many other dogs in the park will share her name. In `PersonPresenter`, you will need to present other data related to a person, like their address or phone number. It's also conceivable that you might split and join other bits of this externally-supplied data. When that day comes, you will be forced into renaming decisions. One ill-advised option would be to use `split_and_join` as a prefix for different kinds of data:
+
+```elixir
+defmodule PersonPresenter do
+  def split_and_join_name(name)
+    name
+    |> String.split()
+    |> Enum.join(" ")
+  end
+
+  def split_and_join_address(address)
+    address
+    |> String.split()
+    |> Enum.join(" ")
+  end
+
+  def split_and_join_phone_number(phone_number)
+    phone_number
+    |> String.split()
+    |> Enum.join(" ")
+  end
+end
+```
+
+Repeated name prefixes (and suffixes) are an indicator for various code smells. Here, the repetition simply boils down to a poorly chosen name. The code above might seem ridiculous on its face, but it's not hard to find function names that mirror implementation details in any sizable codebase. You should reject these kinds of names out of hand. Unless you're writing something similar to a string library, they're not a good idea.
+
+**GUIDELINE: Resist mirroring implementation details in a function name**
+
+### A rung up: eventually false explanations
+
+`split_and_join` describes how the function works. Describing what the function *does* moves the name one level of abstraction higher. You might consider a name like `trim_whitespace`.
+
+```elixir
+defmodule PersonPresenter do
+  def trim_whitespace(name)
+    name
+    |> String.split()
+    |> Enum.join(" ")
+  end
+end
+```
+
+This is better. But recall that the data comes from a source you do not own. It's controlled by someone else and likely to change in unpredictable ways. Even code and data that you *do* own are susceptible to ever-evolving business requirements. The code you write today might seem concrete, but like Molly the dog, even seemingly concrete entities are not static. Code is a being in-process, constantly churning. Your source control repository's diff log is a more accurate mental model of your application than current files on disk.
```

Comment:
It's really OK to start a sentence with "But" occasionally. But you've done it 13 times in this blog post. I might suggest that it's a habit of yours. Maybe aim for three usages, in places where you want a particularly hard stop. Try combining with the previous sentence, "This is better, but recall that the data comes from a source you do not own." Resist the urge to substitute But with However or Actually since they are overused in general.

Reply:
😄 I literally laughed out loud at the second sentence. Wow. Oh my, even. Great catch. I didn't realize I was doing that. I'll go through and implement your suggestions.

Reply:
Addressed in 075a1ac8daa99338b6f2d9cf8a840998173f1678